onnxruntime/core/optimizer/unsqueeze_elimination.cc

```diff
@@ -2,104 +2,108 @@
 // Licensed under the MIT License.
 
 #include "core/optimizer/unsqueeze_elimination.h"
+#include "core/graph/graph_utils.h"
+#include "core/graph/graph.h"
 
 using namespace ONNX_NAMESPACE;
 using namespace ::onnxruntime::common;
 
 namespace onnxruntime {
 
-Status UnsqueezeElimination::ApplyImpl(onnxruntime::Graph& graph, bool& modified, int graph_level) const {
-  std::vector<onnxruntime::NodeIndex> removed_nodes;
-
-  for (auto& node : graph.Nodes()) {
-    // recurse first as there are early exits in the processing here
-    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level));
+Status UnsqueezeElimination::Apply(Graph& graph, Node& node, bool& modified, bool& removed) {
+  // Get "axes" attribute
+  const ONNX_NAMESPACE::AttributeProto* attr = graph_utils::GetNodeAttribute(node, "axes");
+  if (attr == nullptr || attr->type() != AttributeProto_AttributeType_INTS) {
+    return Status::OK();
+  }
```

Comment:
To me, this check should be part of rule and should be put in SatisfyCondition.

Reply:
I added all quick checks in the SatisfyCondition. But axes attribute needs to be accessed in the Apply as well, so we will just be calling it twice.

Reply:
So, that means SatisfyCondition only does partially condition check.

Reply:
Yes, SatisfyCondition tells you that you can go on and apply the rule. It does not say that the rule will succeed.

Reply:
So, the name of SatisfyCondition is kind of confusing here. The people could be confused here when they add new rewrite rules because they need to judge what check needs to be in SatisfyCondition and what doesn't need to.

Reply:
What do you suggest then?
It is a rule-based engine. The point of the rule conditions is so that you fire the rules as few times as possible. However, if you end up duplicating work that you will do in the body of the rule, then your check and the rule as a whole will not be as efficient.
You can put nothing at the SatisfyCondition (and do it all at the body) or you can put all possible checks (and possibly duplicate a lot of work). But you need to strike a balance between the two.

Reply:
I would suggest to add some comments as you described above for API SatisfyCondition. Then, the contributors can refer to these comments to make a trade-off when they are not sure.